### ISSUE TITLE
Error during compile with Mono

### ISSUE THREAD

**Issue by grzegorz-ogrodowski** [cake-build/cake#1528]
Hi, 
I'm trying compile my project using mono (Debian). Project works fine on Windows.
Here I have bad error message:

```
unknown.cake (124,48): Internal compiler error during parsingSystem.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
  at System.ThrowHelper.ThrowArgumentOutOfRangeException (System.ExceptionArgument argument, System.ExceptionResource resource) [0x00011] in <f712f98eb8e445c8918edaf595bbe465>:0 
  at System.ThrowHelper.ThrowArgumentOutOfRangeException () [0x00000] in <f712f98eb8e445c8918edaf595bbe465>:0 
  at System.Collections.Generic.List`1[T].get_Item (System.Int32 index) [0x0000c] in <f712f98eb8e445c8918edaf595bbe465>:0 
  at Mono.CSharp.Location.get_SourceFile () [0x0000f] in <e5610ebfe8c34526a8158db3c89186d4>:0 
  at Mono.CSharp.Location.op_Subtraction (Mono.CSharp.Location loc, System.Int32 columns) [0x00000] in <e5610ebfe8c34526a8158db3c89186d4>:0 
  at Mono.CSharp.Tokenizer.consume_string (System.Boolean quoted) [0x0000f] in <e5610ebfe8c34526a8158db3c89186d4>:0 
  at Mono.CSharp.Tokenizer.xtoken () [0x00d72] in <e5610ebfe8c34526a8158db3c89186d4>:0 
  at Mono.CSharp.Tokenizer.token () [0x00000] in <e5610ebfe8c34526a8158db3c89186d4>:0 
  at Mono.CSharp.CSharpParser.yyparse (Mono.CSharp.yyParser.yyInput yyLex) [0x0010f] in <e5610ebfe8c34526a8158db3c89186d4>:0 
  at Mono.CSharp.CSharpParser.parse () [0x0002e] in <e5610ebfe8c34526a8158db3c89186d4>:0 
```

I'm using build.sh downloaded from this project..

Mono version: 4.8.0
OS version: Debian GNU/Linux 8 (jessie)

**Comment by gep13** [comment]
@grzegorz-ogrodowski is it possible for you to share the script that you are trying to compile?

**Comment by grzegorz-ogrodowski** [comment]
It is a commercial project and I'm affraid that I can't do it for you.

**Comment by gep13** [comment]
@grzegorz-ogrodowski you understand that without a reproducible example of the problem, it is going to be very hard to help you, right?

**Comment by grzegorz-ogrodowski** [comment]
Yes, I understand. But I had hoped that is quite known issue.
It's really big project created in ASP NET CORE. 
I'm using build.sh script from this project.

If this is unacceptable - I'll close this issue ASAP.

**Comment by gep13** [comment]
@grzegorz-ogrodowski it is not that it is not acceptable, it's just that there is very little that we can do unless we are able to reproduce the problem.  Are you able to try to reproduce the problem on a smaller project, and share that with us?

**Comment by devlead** [comment]
It's hard without the details, one thing you could ensure is that line endings are unix style, can usually be controlled through I.e. Gitattributes.

Also you could start with an empty cake file and methodically add stuff until it fails.

A more wild option would be too switch to the Cake.CoreCLR runner which uses the same scripting engine regardless of platform, drawback is that requires .NET Core installed and any addins used to be .NET compatible.

**Comment by grzegorz-ogrodowski** [comment]
After few hours of investigation I found that my nuget has problem with certificate (even if CA is installed).
Maybe this is problem. I'll fix my nuget and then back to cake+mono.

**Comment by gep13** [comment]
Sounds good. Let us know how you get on.

**Comment by ErikSchierboom** [comment]
@grzegorz-ogrodowski Were you able to fix this? I also have the same problem.

**Comment by grzegorz-ogrodowski** [comment]
@ErikSchierboom unfortunately not. I have still a lot troubles with nuget and CA issue.

**Comment by Ashthos** [comment]
Check for syntax errors, things like missing semi colons. 

Or check things like auto initialisers `var thing = { "string" };` as mono can't cope with these and will report an error similar to the one you posted. 

Note: Windows can cope with the things i've detailed fine.

**Comment by ErikSchierboom** [comment]
In my script, it indeed turned out to be a missing semicolon.

**Comment by patriksvensson** [comment]
Closed since #1645 been merged to develop. Feel free to reopen this if it's still a problem after 0.22.0 have been released.
